mesonbuild/interpreterbase/interpreterbase.py

```diff
@@ -783,19 +783,11 @@ def array_method_call(self,
                           posargs: T.List[TYPE_var],
                           kwargs: TYPE_kwargs) -> T.Union[TYPE_var, InterpreterObject]:
         if method_name == 'contains':
-            def check_contains(el: list) -> bool:
-                if len(posargs) != 1:
-                    raise InterpreterException('Contains method takes exactly one argument.')
-                item = posargs[0]
-                for element in el:
-                    if isinstance(element, list):
-                        found = check_contains(element)
-                        if found:
-                            return True
-                    if element == item:
-                        return True
-                return False
-            return check_contains(obj)
+            elements = mesonlib.listify(_unholder(obj), flatten=True)
+            posargs = mesonlib.listify(posargs, flatten=True)
+            if len(posargs) != 1:
+                raise InterpreterException('Contains method takes exactly one argument.')
```

Comment:
This actually breaks this:

```meson
list = [['a', 'b'], ['c', 'd']]
message(list.contains(['a', 'b']))
```

Returns `true` in 0.58 and `true` in master, but causes a hard error with this PR.

This change also breaks this:

```meson
flist = [files('a.c', 'b.c'), files('foo.c', 'bar.c')]
message(flist.contains(files('a.c', 'b.c')))
```

Returns `true` in 0.58 and `false` in master (incorrectly), but with this change it causes a hard error.

I think the correct fix is not to mangle `posargs` but to check for both `obj` and `_unholder(obj)`. That does actually work for me:

```diff
@@ -787,13 +787,14 @@ The result of this is undefined and will become a hard error in a future Meson r
                 if len(posargs) != 1:
                     raise InterpreterException('Contains method takes exactly one argument.')
                 item = posargs[0]
-                for element in el:
-                    if isinstance(element, list):
-                        found = check_contains(element)
-                        if found:
+                for each in (el, _unholder(el)):
+                    for element in each:
+                        if isinstance(element, list):
+                            found = check_contains(element)
+                            if found:
+                                return True
+                        if element == item:
                             return True
-                    if element == item:
-                        return True
                 return False
             return check_contains(obj)
         elif method_name == 'length':
```

Reply:
Note that this doesn't fix the two cases below that didn't work in 0.58, so it likely needs more changes.

Reply:
I think `listify` is wrong for generic array operations. It's only correct when passing args to things that expect a flattened list, like files list for a target or a command-line argument list.

Reply:
I dropped the `listify` refactoring and reverted back to one `_unholder` pass at the start.